search-provider/yelp-search-provider.c

```diff
@@ -0,0 +1,603 @@
+        g_error_free (error);
+        return;
+    }
+    g_dbus_method_invocation_return_value (invocation, NULL);
+    return;
+}
+
+static gboolean 
+handle_launch_search (YelpShellSearchProvider2  *skeleton,
+                      GDBusMethodInvocation     *invocation,
+                      gchar                    **terms,
+                      guint32                    timestamp,
+                      gpointer                   user_data)
+{
+    YelpSearchProviderApp *app = YELP_SEARCH_PROVIDER_APP (user_data);
+    gchar *joined_terms = g_strjoinv (" ", terms);
+
+    yelp_launch (joined_terms, app, invocation, "show-search");
+
+    g_free (joined_terms);
+
+    return TRUE;
+}
+
+static gboolean
+handle_activate_result (YelpShellSearchProvider2  *skeleton,
+                        GDBusMethodInvocation     *invocation,
+                        gchar                     *result,
+                        gchar                    **terms,
+                        guint32                    timestamp,
+                        gpointer                   user_data)
+{
+    YelpSearchProviderApp *app = YELP_SEARCH_PROVIDER_APP (user_data);
+
+    yelp_launch (result, app, invocation, "show-page");
+
+    return TRUE;
+}
+
+static void
+search_provider_app_dispose (GObject *obj)
+{
+    YelpSearchProviderApp *self = YELP_SEARCH_PROVIDER_APP (obj);
+
+    g_clear_object (&self->skeleton);
+    g_clear_pointer (&self->page_data_hash_map, g_hash_table_unref);
+
+    G_OBJECT_CLASS (yelp_search_provider_app_parent_class)->dispose (obj);
+}
+
+static void
+preload_data_cb (YelpDocument          *document,
+                 YelpDocumentSignal     signal,
+                 YelpSearchProviderApp *app,
+                 GError                *error)
+{
+    YelpSearchProviderApp *self = YELP_SEARCH_PROVIDER_APP (app);
+    gchar **page_ids = yelp_document_list_page_ids (document);
+    gchar **iter;
+
+    if (signal == YELP_DOCUMENT_SIGNAL_ERROR) {
+        g_warning ("error during preloading data: %s", (error != NULL) ? error->message : "unknown error");
+        return;
+    }
+
+    if (signal == YELP_DOCUMENT_SIGNAL_CONTENTS)
+        return;
+
+    if (signal == YELP_DOCUMENT_SIGNAL_INFO) {
+        for (iter = page_ids; *iter; iter++) {
+            gchar *page_id = *iter;
+            PageData *data = g_new0 (PageData, 1);
+            gchar *icon_string = yelp_document_get_page_icon (document, page_id);
+            GIcon *icon = g_icon_new_for_string (icon_string, NULL);
+            g_free (icon_string);
+
+            data = page_data_new_steal (yelp_document_get_page_title (document, page_id), 
+                                        yelp_document_get_page_desc (document, page_id), 
+                                        icon);
+            g_hash_table_insert (app->page_data_hash_map, page_id, data);
+        }
+
+        g_application_release (G_APPLICATION (self));
```

Comment:
After reading the yelp code I think that the only reason you get the SIGNAL_ERROR is that the requested page id was not found, so SIGNAL_INFO will not happen. Then yes, I think that after getting SIGNAL_ERROR we should release the application before returning from the function.

Since the semantics of the error signal are not documented then I don't know if it might be possible to get an error because of some syntax error in the document, so we would end up with receiving SIGNAL_INFO and then SIGNAL_ERROR. I guess we would need to guard ourselves against releasing the app too many times with some boolean member in the app.